### PR TITLE
Support using AsyncProducer

### PR DESCRIPTION
**Pull request by nullaus**
• Added config options for both AsyncProducer (auto delivery options) and Producer (buffer size options)
• Use 'logger' from WaterDrop::logger to get errors logged when occurring in AsyncProducer
• Updated tests

### CODE REVIEW

File: lib/water_drop/producer_proxy.rb [hunk 1]
```diff
@@ -46,6 +47,11 @@ def send_message(message)
       @attempts = 0
     end
 
+    def shutdown
+      @producer.shutdown if @producer
+      @producer = nil
```
Comment (nullaus) [inline lib/water_drop/producer_proxy.rb]: This could potentially call 'reload!'.

**Comment by mensfeld** [comment]
Hey @nullaus  could you describe why you need Waterdrop in an async mode and your use-case for that??

Waterdrop was meant to be a simple library that could be used with Karafka for responders (or without but in a simple way). Merging this will give us certain "business" benefits but on the other hand I'm a bit afraid. Logic behind async Kafka producer in Ruby-Kafka is really, really, really complex and it involves handling separate threads, their communication, errors, logging, etc? Did you consider in your code, that we maintain connection pool for Waterdrop? Based on your code I think that you didn't and this would mean that for each producer Waterdrop will maintain a separate "backend" including threads and all the heavy stuff. I don't think, that this should work that way.

If you plan to use it with Karafkas responders, maybe it would be better to create a responders adapter layer and hook up directly the Ruby-kafka gem into Responders flow (we can do that)? Adding this to Waterdrop without a really deep testing feels risky.

@mach-kernel @agwozdowski @pavlo-vavruk what do you think?

**Comment by nullaus** [comment]
@mensfeld Sure thing. The reason we need it is that submitting jobs to Kafka can get slow with the amount of traffic we get, and the amount of jobs we create. To alleviate this, we added async submission. We don't actually respond to the job output. In this particular case, for us, the workers handle the errors, retries, etc. I would certainly suggest that to people that would use this asynchronous submission option.

The logger being set how it is will output all of the same information as it does in synchronous submissions. It's what I used to debug parts of it.

Happy to provide more information, and apologies for the slow response.

**Comment by mensfeld** [comment]
Hey @nullaus really sorry for the delay. I'm back now :)

Thank you for your description, now I understand your context and I have few questions that I would appreciate if you could answer:

1. did you consider the impact of connection pooling pool on the async producers threads performance?
2. did you consider changing (in the sync producer version) the ack level to improve the performance?
3. what about flushing the remaining messages on shutdown (in each pool message producer instance)?

The way waterdrop works is by managing the connection pool of producents that is a must be for an sync producer. But it is not true for the async one. We could add this functionality but we would have to agree on two things:

1. Do we use connection pooling for async (which I think is bad) or do we use a single producer (it should be thread safe as it works with multi threaded environments (and it is recommended by ruby-kafka)?
2. What do we do with the shutdown process.

Once again, really sorry for the delay. I should be more active now :-)

**Comment by mensfeld** [comment]
@nullaus any update on that one?

**Comment by nullaus** [comment]
1. Yes, and for us, slightly slowed job bootstrapping is better than slower submission.
2. Yes, but it doesn't improve the performance enough for our needs.
3. It should currently allow you to flush as long the entire process tree doesn't crash (which would mean the call to drain and shutdown would not happen)
4. Yes, we're aware it may not drain completely and for us that's an acceptable loss of data (but we could probably fix this)

Follow-up questions.

1. I didn't want to pry too much into how things were currently built but I would imagine that simply having a single producer for async would be fine. It would save some resources for sure. Consider my implementation to be more of an MVP than a fully fleshed out implementation. :)
2. Good question! Like anything async, even with draining and waiting we may find ourselves unable to send everything if it takes too long -- we'll likely get killed to finish shutting down within a reasonable amount of time. I'd instead go for a two pronged approach, where depending on the queue size, we would either flush and shutdown OR serialize to disk so that upon restart we can submit the rest (as a configuration option, since for some these jobs would be considered stale and useless)

My apologies too for such a slow reply. I completely forgot about this thread. :)

**Comment by filiptepper** [comment]
Bumping this one, just in case. :)

**Comment by mensfeld** [comment]
@filiptepper  no need to. This is a complex thing that will be resolved once I have this one: https://github.com/zendesk/ruby-kafka/issues/424 as I find manual offset commits more important

**Comment by filiptepper** [comment]
Sounds good!

**Comment by mensfeld** [comment]
@filiptepper @nullaus  waiting for some fixed in ruby-kafka, got tnis in progress. Looks promising but due to buffering, requires new karafka release as well. Work in progress. Probably in 2-3 weeks will be done :)

**Comment by nullaus** [comment]
Sweet! Thanks for the status update. :)

**Comment by mensfeld** [comment]
Async is redesigned so unfortunately this won't be needed but thank you for your contribution. Feel free to review #41

**Pull request closed without merging**